[SHK-28] Coupon 발급 기능 구현

goseungwon opened the pull request:
### ⛏ 작업 사항
- Coupon Entity 작성
- CouponEvent Entity 작성
- Coupon 생성 API 작성
- Coupon 발급 API 작성
- Coupon 생성 API Test
- Coupon 발급 API Test

### 📝 작업 요약
- 쿠폰 생성 기능 및 테스트 코드 작성
- 쿠폰 발급 기능 및 테스트 코드 작성


### 💡 관련 이슈
- Resolved : [SHK-108], [SHK-97], [SHK-119], [SHK-37], [SHK-129]


[SHK-108]: https://shoekream.atlassian.net/browse/SHK-108?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SHK-97]: https://shoekream.atlassian.net/browse/SHK-97?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SHK-119]: https://shoekream.atlassian.net/browse/SHK-119?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SHK-37]: https://shoekream.atlassian.net/browse/SHK-37?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SHK-129]: https://shoekream.atlassian.net/browse/SHK-129?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

goseungwon committed:
Feat : [SHK-108] 쿠폰 entity 생성

goseungwon committed:
Feat : [SHK-97] 쿠폰 등록 API 제작

goseungwon committed:
Merge remote-tracking branch 'origin/feature/SHK-28' into feature/SHK-28

goseungwon committed:
Feat : [SHK-37] 쿠폰 발급 API 제작

변경 내용
1. CouponService 메서드명 변경

goseungwon committed:
Merge remote-tracking branch 'origin/develop' into feature/SHK-28

goseungwon committed:
Remove : Dto 수정

변경 내용
1. CouponControllerResponse 삭제
2. CouponServiceResponse -> CouponRegisterResponse 이름 변경

변경 이유
1. CouponServiceResponse 와 중복됨
2. 새로운 이름이 더 적절하다 판단

goseungwon committed:
Test : [SHK-119] 쿠폰 등록 API 테스트

- CouponServiceTest 구현
- CouponControllerTest 구현

변경 내용
1. Coupon 생성자에 id 추가

변경 이유
1. TestEntity 생성에 제한됨

goseungwon committed:
Test : [SHK-129] 쿠폰 발급 API 테스트

- CouponEventServiceTest 구현
- CouponEventControllerTest 구현

변경 내용
1. CouponEvent 생성자에 id 추가

변경 이유
1. TestEntity 생성에 제한됨

goseungwon committed:
Chore : Dto 컨벤션 변경으로 인한 수정

변경 내용
1. dto 패키지 밑에 request, response 로 분리

변경 이유
1. 팀 컨밴션 변경

boompatron commented on src/main/java/com/prgrms/kream/domain/coupon/controller/CouponController.java:
```diff
@@ -1,4 +1,43 @@
+	 * @param couponRegisterRequest 할인율, 쿠폰 이름, 쿠폰 수량
+	 * @return CouponControllerResponse
+	 * @see CouponService
+	 */
```
Java Doc을 적용하셨군요! 저도 적용해야 할 것같네요!!

boompatron reviewed (approved):
구현하느라 고생하셨습니다!

BnDC commented on src/main/java/com/prgrms/kream/common/mapper/CouponEventMapper.java:
```diff
@@ -0,0 +1,25 @@
+	private CouponEventMapper() {
+	}
+
+	public static CouponEvent toCouponEvent(CouponEventServiceRequest couponEventServiceResponse) {
```
`couponEventServiceResponse` -> `couponEventServiceRequest`로 수정해주세요~

BnDC commented on src/main/java/com/prgrms/kream/common/mapper/CouponEventMapper.java:
```diff
@@ -0,0 +1,25 @@
+				.build();
+	}
+
+	public static CouponEventDto toCouponEventControllerResponse(CouponEvent couponEvent) {
```
`CouponEventDto`와 `toCouponEventControllerResponse` 를 `CouponEventResponse` 와 `toCouponEventResponse`로 통일하면 좋을 거 같습니다.

BnDC commented on src/main/java/com/prgrms/kream/common/mapper/CouponMapper.java:
```diff
@@ -1,4 +1,26 @@
+import com.prgrms.kream.domain.coupon.model.Coupon;
+
 public class CouponMapper {
+	private CouponMapper() {
+	}
```
`@NoArgsConstructor(access=PRIVATE)`어노테이션이 아닌 직접 작성하신 이유가 있을까요?

goseungwon replied:
NoArgsConstructor 의 속성을 활용할 생각을 하지 못했습니다

BnDC commented on src/main/java/com/prgrms/kream/common/mapper/CouponMapper.java:
```diff
@@ -1,4 +1,26 @@
+				.build();
+	}
+
+	public static CouponDto toCouponRegisterResponse(Coupon coupon) {
```
이것도 method이름과 반환타입을 통일하면 좋을거 같습니다

BnDC commented on src/main/java/com/prgrms/kream/domain/coupon/controller/CouponController.java:
```diff
@@ -1,4 +1,43 @@
+	public ApiResponse<CouponDto> couponRegister(
+			@RequestBody @Valid CouponRegisterRequest couponRegisterRequest
+	) {
+		CouponDto couponRegisterResponse = couponService.registerCoupon(couponRegisterRequest);
+
+		return ApiResponse.of(couponRegisterResponse);
```
`couponRegisterResponse` 지역 변수를 만드는 이유가 궁금합니다

goseungwon replied:
Q. 전체적으로 지역변수를 만들고 다른 처리를 하지 않고 리턴하는 구문이 많은데 지역변수를 만드는 이유가 있을까요?
A. 그렇게 한 이유는 가독성이 좋다고 생각해서 입니다. 이 부분은 토론해보면 좋을 것 같아요.

BnDC commented on src/main/java/com/prgrms/kream/domain/coupon/dto/request/CouponEventRegisterRequest.java:
```diff
@@ -0,0 +1,11 @@
+
+public record CouponEventRegisterRequest(
+		@NotNull(message = "쿠폰은 필수 입력사항 입니다.")
+		long coupon_id,
+		@NotNull(message = "멤버는 필수 입력사항 입니다.")
+		long member_id
```
이 부분은 camelCase인 이유가 있을까요?

goseungwon replied:
제 실수입니다..! 찾아주셔서 감사합니다 🌞

BnDC reviewed (commented):
기능 구현하느라 고생하셨습니다~ 몇가지 코멘트 확인해주면 좋을 거 같아요

코멘트는 하나만 남겼지만, 전체적으로 지역변수를 만들고 다른 처리를 하지 않고 리턴하는 구문이 많은데 지역변수를 만드는 이유가 있을까요?

goseungwon committed:
Refactor : pr 리뷰 반영

변경 내용
1. CouponMapper/CouponEventMapper 메서드 불일치 수정
2. CouponMapper/CouponEventMapper 생성자 어노테이션으로 변경
3. CouponEventRegisterRequest 필드명 변경

변경 이유
1. 반환값과 메서드명 불일치
2. 가독성을 증진하기 위함
3. 컨벤션 위배

BnDC commented on src/main/java/com/prgrms/kream/domain/coupon/service/CouponEventService.java:
```diff
@@ -16,11 +16,11 @@
 	private final CouponEventRepository couponEventRepository;
 
-	public CouponEventDto registerCouponEvent(CouponEventServiceRequest couponEventServiceResponse) {
+	public CouponEventResponse registerCouponEvent(CouponEventServiceRequest couponEventServiceResponse) {
```
이 부분도 `CouponEventServiceRequest couponEventServiceResponse` 부분의 변수명을 통일하면 좋을거 같아요

goseungwon replied:
이 부분은 고민을 했었는데, 공통 dto로 계속 사용될 것이라 판단되어 네이밍을 저렇게 했습니다.
어떻게 생각하시나요?

goseungwon committed:
Refactor : pr 리뷰 반영

변경 내용
1. registerCouponEvent 인자 수정
2. Controller 변수명 변경

변경 이유
1. 타입과 메서드명 불일치
2. 타입과 메서드명 불일치

goseungwon committed:
Refactor : CouponFacade 에서 Entity 대신 Dto 사용

변경 내용
1. Coupon -> CouponResponse

변경 이유
1. Entity 의존성 제거
